subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c

```diff
@@ -276,12 +276,12 @@ static int mbox_init(const struct device *instance)
 
 	k_work_init(&data->mbox_work, mbox_callback_process);
 
-	err = mbox_register_callback(&conf->mbox_rx, mbox_callback, data);
+	err = mbox_set_enabled(&conf->mbox_rx, true);
 	if (err != 0) {
 		return err;
 	}
 
-	return mbox_set_enabled(&conf->mbox_rx, 1);
+	return mbox_register_callback(&conf->mbox_rx, mbox_callback, data);
```

Comment:
What happen in case the remote processor (running as rpmsg host) sends a first message before this function is called?
Calling mbox_set_enabled  means that pending signal are cleaned? I would expect instead that a pending signals would be treated.
In this case no callback is registered on enable, the signal  is dropped.
So i wonder if the issue should not be fixed in the mailbox instead of creating potential side effect here.

Reply:
> What happen in case the remote processor (running as rpmsg host) sends a first message before this function is called?

I don't think that's possible. Before sending a first message, both host and remote must have gone thrugh the `bound()` callback and that means that both have gone through this function already.

> Calling mbox_set_enabled means that pending signal are cleaned? I would expect instead that a pending signals would be treated.

Before `mbox_set_enabled()` you are not supposed to receive any signal at all.

> So i wonder if the issue should not be fixed in the mailbox instead of creating potential side effect here.

I disagree here. The problem is that you cannot install a callback on an unused (not enabled) channel. To be able to install a callback you have to enable a channel and later install a callback on that channel. If a signal arrives between `set_enable` and `register_callback` this is simply discarded. At least this is what I had in mind when I planned this API.

If we want to do the opposite (that is you can install a callback on a non-enabled channel and you must enabled a channel on which you have already installed a callback) that means that I have to modify the current MBOX IPC driver. But given that I don't think this solution gives us any real advantage I would prefer to modify the backend instead of MBOX API and driver.

Reply:
>> What happen in case the remote processor (running as rpmsg host) sends a first message before this function is called?

> I don't think that's possible. Before sending a first message, both host and remote must have gone thrugh the `bound()` callback and that means that both have gone through this function already.

The first message sent is the Name Service anoucement on endpoint creation. A rpmsg host can configure the vrings and create its endpoint, that will generate an mbox signal.

>> Calling mbox_set_enabled means that pending signal are cleaned? I would expect instead that a pending signals would be treated.

> Before `mbox_set_enabled()` you are not supposed to receive any signal at all.

An example is the load of a co-processor firmware by Linux. it sends a mailbox signal to the coprocessor when it has initialised the vrings. This signal can be set befor the Zephyr initialises its mailbox. 
More than that the mailbox can be used for some other purpose that the rpmsg communication. Dropping a signal can be a problem. So at least having a sequence that handles the use case seems safer to me.

>>  So i wonder if the issue should not be fixed in the mailbox instead of creating potential side effect here.

> I disagree here. The problem is that you cannot install a callback on an unused (not enabled) channel. To be able to install a callback you have to enable a channel and later install a callback on that channel. If a signal arrives between `set_enable` and `register_callback` this is simply discarded. At least this is what I had in mind when I planned this API.

This is the key point. For me as you declare the phandle in the DT the channel is used. `mbox_set_enabled` should not be used to declare the use of the channel but only to enable it. 
In fact, I would expect to have the same behaviour that the IRQ mechanism. Indeed enable an IRQ management without registering an handler in the vector table generally leads to a crash.

I also have in mind the suspend and resume mechanismes to handle the low power. We can have a use case where the mailbox is used for signaling (no rpmsg). Before going into standby, we could want to disable the channel and re-enable it at resume time. So this means that we should also re-register the callback?

In other world the `mbox_register_callback` should be part of the initialisation API, while the `mbox_set_enabled` seems more part of the runtime API.

> If we want to do the opposite (that is you can install a callback on a non-enabled channel and you must enabled a channel on which you have already installed a callback) that means that I have to modify the current MBOX IPC driver. But given that I don't think this solution gives us any real advantage I would prefer to modify the backend instead of MBOX API and driver.

Could you detail what would have to be updated in MBOX API? 
Perhaps the first step is to define the more generic API: register the callback before enabling the channel or the opposite?

Reply:
> The first message sent is the Name Service anoucement on endpoint creation. A rpmsg host can configure the vrings and create its endpoint, that will generate an mbox signal.

With the current backend I don't think it is possible. When the host creates the endpoint if this is not already been advertised by the remote core, it is simply cached (nothing happens), waiting for the remote core announcement. The first to advertise the endpoints is always the remote endpoint in the current implementation.

>This is the key point. For me as you declare the phandle in the DT the channel is used. mbox_set_enabled should not be used to declare the use of the channel but only to enable it.
In fact, I would expect to have the same behaviour that the IRQ mechanism. Indeed enable an IRQ management without registering an handler in the vector table generally leads to a crash.
> ...
> In other world the mbox_register_callback should be part of the initialisation API, while the mbox_set_enabled seems more part of the runtime API.

Ok, I understand the use case and your points. I'll change the API and adapt the current driver.

> Could you detail what would have to be updated in MBOX API?

Mostly documentation. But I have to change the (only) MBOX driver to be compliant.

> Perhaps the first step is to define the more generic API: register the callback before enabling the channel or the opposite?

At this point I'm proposing to change the API to force the user to register the callback before enabling the channel. I'll prepare a PR for that and I'll close this one.